crates/rust-client/src/sync/state_sync.rs

```diff
@@ -142,30 +144,58 @@ impl StateSync {
             ..Default::default()
         };
 
-        while self
-            .sync_state_step(
+        let current_block_num = (u32::try_from(current_partial_mmr.num_leaves() - 1)
+            .expect("The number of leaves in the MMR should be greater than 0 and less than 2^32"))
+        .into();
+        let account_ids: Vec<AccountId> = accounts.iter().map(AccountHeader::id).collect();
+
+        let mut stream =
+            self.rpc_api.sync_state(current_block_num, &account_ids, &note_tags).await?;
+
+        while let Some(res) = stream
+            .message()
+            .await
+            .map_err(|e| ClientError::RpcError(ConnectionError(e.message().to_string())))?
+        {
+            self.sync_state_step(
+                res.try_into().map_err(ClientError::RpcError)?,
                 &mut state_sync_update,
                 &mut current_partial_mmr,
                 &accounts,
-                &note_tags,
-                &nullifiers_tags,
             )
-            .await?
-        {
-            // New nullfiers should be added for new untracked notes that were added in previous
-            // steps
-            nullifiers_tags.append(
-                &mut state_sync_update
+            .await?;
+        }
+
+        // New nullfiers should be added for new untracked notes
+        nullifiers_tags.append(
+            &mut state_sync_update
+                .note_updates
+                .updated_input_notes()
+                .map(|note| get_nullifier_prefix(&note.nullifier()))
+                .collect::<Vec<_>>(),
+        );
+        let new_nullifiers = self
+            .rpc_api
+            .check_nullifiers_by_prefix(&nullifiers_tags, current_block_num)
+            .await?;
+
+        // Process nullifiers
+        let transactions = vec![];
+        let mut discarded_transactions = vec![];
+        for (nullifier, block_num) in new_nullifiers {
+            let nullifier_update = NullifierUpdate { nullifier, block_num };
+            if (self.on_nullifier_received)(nullifier_update.clone()).await? {
+                let discarded_transaction = state_sync_update
                     .note_updates
-                    .updated_input_notes()
-                    .filter(|note| {
-                        note.is_committed()
-                            && !nullifiers_tags.contains(&get_nullifier_prefix(&note.nullifier()))
-                    })
-                    .map(|note| get_nullifier_prefix(&note.nullifier()))
-                    .collect::<Vec<_>>(),
-            );
+                    .apply_nullifiers_state_transitions(&nullifier_update, &transactions)?;
+
+                if let Some(transaction_id) = discarded_transaction {
+                    discarded_transactions.push(transaction_id);
+                }
+            }
         }
+        let transaction_updates = TransactionUpdates::new(transactions, discarded_transactions);
```

Comment:
I would remove `transactions` if we don't use it:
```suggestion
        let transaction_updates = TransactionUpdates::new(vec![], discarded_transactions);
```